packages/salesforcedx-apex-debugger/src/adapter/apexDebug.ts

```diff
@@ -460,6 +467,9 @@ export class ApexDebug extends LoggingDebugSession {
   private trace: string[] | undefined;
   private traceAll = false;
 
+  private lock = new AsyncLock({ timeout: DEFAULT_LOCK_TIMEOUT_MS });
```

Comment:
This seems to imply that we have a lock for each MapReferenceContainer? Is the MapReferenceContainer a singleton? It doesn't have the word service or singleton in it so I'm not sure. 

I'm asking because I want to know how many locks are we creating.

If we don't have a shared lock, then we cannot actually _serialize_ the operations and we might still end up with a race condition.

Reply:
Also, just to be certain, all the requests for Create, Update, Delete, And Read (read is important as well or else we can have a race condition if on the data) go through this entry point?

Reply:
This lock belongs to the ApexDebug object. VS Code ensures there can only be one 'apex' debug type active at a time, so there will only be one lock. All breakpoint operations go through setBreakpointsRequest where the lock is used.